### ISSUE TITLE
[Parity 10/16] GPU primitives breadth - cuDNN RNN, cuSPARSE, cuSOLVER, cuRAND, Thrust/CUB, NVTX, mem snapshot

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#219]
Part of epic #209.

## Summary

We have cuDNN Conv + BN, cuBLAS + cuBLASLt, NCCL. The rest of PyTorch's CUDA primitive surface is missing: cuDNN RNN/LSTM/GRU, cuSPARSE, cuSOLVER, cuRAND (GPU-native RNG), cuFFT (tracked by FFT issue), Thrust/CUB (reduce/scan/sort), NVTX ranges, and memory-snapshot APIs. This also covers corresponding primitives on HIP/Metal/OpenCL/Vulkan where they exist.

### cuDNN
- RNN / LSTM / GRU (uni/bi-directional, multi-layer, dropout between layers, projections, persistent kernels)
- DropoutDescriptor + fused dropout in RNN
- Attention + multi-head attention primitives (newer cuDNN)

### cuSPARSE
- `sparse_coo_tensor.matmul(dense)`, `sparse_csr_tensor.matmul(dense)`
- SpMM, SpMV, SpGEMM
- Sparse-sparse operations

### cuSOLVER (tracked mostly in linalg issue; GPU bindings live here)
- `potrf` / `potrs` (Cholesky)
- `geqrf` / `orgqr` (QR)
- `gesvdj` / `gesvd` (SVD — we have partial)
- `syevd` / `syevj` (symmetric eigen)
- `getrf` / `getrs` (LU + solve)

### cuRAND
- Philox / MRG32k3a / Sobol / Scrambled Sobol on-device RNG
- Device-side streams keyed by `(seed, subsequence, offset)`

### Thrust / CUB
- Device-wide reduce, scan, sort, radix-sort, run-length encode, histogram
- Warp-level / block-level primitives exposed for custom kernels

### NVTX / debug
- `nvtx.range_push`, `range_pop`, `mark`
- `torch.cuda.memory_snapshot()` — allocator state + allocation stacks
- `torch.cuda.memory_summary()`, `memory_stats()`, `reset_peak_memory_stats()`
- `torch.cuda.nvtx.range` context manager

## Current state in the repo

- cuDNN: Conv + BN only (confirmed by `CuDnnNative.cs`)
- cuBLAS / cuBLASLt: fused-epilogue matmul (Wave 1 ✅)
- NCCL: ✅
- cuSPARSE / cuSOLVER / cuRAND / Thrust / CUB / NVTX: ❌
- `GpuWorkspace.cs`, `NativeMemoryManager.cs` handle allocation but no public snapshot API.

## How we beat PyTorch

1. **cuDNN RNN with our symbolic backward**: `SymbolicBackwardGraph` + algebraic simplification means gradients through LSTM states can be simplified at compile time — PyTorch runs them numerically.
2. **cuSPARSE SpMM driven by `AutotuneCache`**: measure the tipping point between dense GEMM and SpMM per sparsity ratio; autotune picks. PyTorch picks statically.
3. **cuRAND exposed as `IPhiloxGenerator<T>`**: same interface as our CPU Philox (see distributions issue) so deterministic seeding works identically on CPU and GPU. Unique — PyTorch's CPU and GPU generators differ.
4. **Thrust/CUB primitives as `Engines/DirectGpu/CUDA/DevicePrimitives.cs`** exposed at tensor level: `DeviceReduce`, `DeviceScan`, `DeviceSort`, `DeviceRunLengthEncode`. PyTorch hides these behind tensor ops; we surface them as a building block so custom CUDA kernels from users can reuse our warp-friendly implementations.
5. **NVTX ranges auto-emitted from `TapeStepContext` and `LazyTensorScope`** — every forward step shows up as a named NVTX range in Nsight Systems without manual annotation. PyTorch requires `torch.autograd.profiler.record_function` calls.
6. **Memory snapshot that includes `TensorArena` + `GpuWorkspace`**: unified view across all our allocators. PyTorch's snapshot only covers the caching allocator.
7. **Multi-backend equivalents where they exist**: rocSPARSE / rocSOLVER / rocRAND for HIP; Metal Performance Shaders for Cholesky/QR/SVD/RNG; Vulkan kernels for scan/reduce; WebGPU kernels. Extends the primitive surface to backends PyTorch doesn't cover.
8. **Persistent RNN kernels** from cuDNN — PyTorch turned this off by default due to bugs; we ship it with correctness tests.

### CUDA
- [ ] cuDNN RNN / LSTM / GRU (uni/bi, multi-layer, dropout, projection, persistent)
- [ ] cuSPARSE bindings + `SparseTensor.MatMul(dense)` + `SpGEMM`
- [ ] cuSOLVER bindings (Cholesky, QR, SVD, eigen, LU) — used by linalg issue
- [ ] cuRAND bindings + `IPhiloxGenerator<T>` GPU impl + Sobol / Scrambled Sobol
- [ ] Thrust/CUB wrappers: `DeviceReduce`, `DeviceScan`, `DeviceSort`, `DeviceRadixSort`, `DeviceHistogram`, `DeviceRunLengthEncode`
- [ ] NVTX ranges (`Nvtx.Range`, `Nvtx.Mark`)
- [ ] `CudaMemory.Snapshot()`, `Summary()`, `Stats()`, `ResetPeak()`
- [ ] Auto-NVTX from `LazyTensorScope` + `TapeStepContext`

### HIP / ROCm
- [ ] MIOpen RNN bindings
- [ ] rocSPARSE / rocSOLVER / rocRAND wrappers
- [ ] ROCm Thrust equivalent

### Metal
- [ ] MPS matrix decompositions
- [ ] Metal RNG
- [ ] Metal Performance Shaders for vision primitives (informs vision issue)

### OpenCL / Vulkan / WebGPU
- [ ] Device-wide reduce, scan, sort kernels (baseline primitives)
- [ ] Device RNG (xorshift-derived where no vendor lib exists)

### Unified surface
- [ ] `IDeviceRng<T>` abstraction with backend-specific implementations
- [ ] `IDevicePrimitives<T>` abstraction for reduce/scan/sort

## Acceptance criteria

- cuDNN RNN: LSTM train-loop on a language-model mini-benchmark matches CPU reference within tolerance.
- cuSPARSE: SpMM correctness test vs dense reference at ≥90% sparsity.
- cuRAND: deterministic test — same `(seed, subsequence, offset)` produces identical bits on CPU Philox and cuRAND Philox.
- Memory snapshot: snapshot during a training step lists all `TensorArena` + `GpuWorkspace` allocations with correct sizes.
- NVTX ranges visible in Nsight Systems without user annotation.
- Benchmarks: LSTM forward/backward throughput ≥ PyTorch + cuDNN; SpMM auto-dispatch beats always-dense on sparse Transformers.

## Implementation notes

- `Engines/DirectGpu/CUDA/CuDnnRnn*.cs`, `CuSparseNative.cs` + `CuSparse.cs`, `CuSolverNative.cs` + `CuSolver.cs`, `CuRandNative.cs` + `CuRand.cs`.
- `Engines/DirectGpu/CUDA/DevicePrimitives.cs` wraps Thrust/CUB (via a small C++ shim if needed).
- `Engines/DirectGpu/CUDA/Nvtx.cs` — pure P/Invoke.
- Snapshot API lives in `Helpers/NativeMemoryManager.cs` + `Helpers/GpuWorkspace.cs`.
- Each HIP / Metal / OpenCL / Vulkan / WebGPU backend gets equivalent files under its respective folder.

**Comment by ooples** [comment]
# Implementation plan — detailed

## 1. Current state (file-level)

| Component | Status |
|---|---|
| cuDNN Conv + BN | ✓ via `CuDnnNative.cs` |
| cuBLAS + cuBLASLt | ✓ Wave 1 (fused-epilogue matmul) |
| NCCL | ✓ via `NcclNative.cs` + `NcclComm.cs` |
| cuSPARSE / cuSOLVER / cuRAND / Thrust / CUB / NVTX | ❌ |
| `GpuWorkspace.cs`, `NativeMemoryManager.cs` | allocation ✓; no public snapshot API |

## 2. Scope — full inventory

### CUDA
```
cuDNN RNN / LSTM / GRU (uni/bi, multi-layer, dropout, projection, persistent)
cuSPARSE bindings + SparseTensor.MatMul(dense) + SpGEMM (feeds #221)
cuSOLVER bindings (Cholesky, QR, SVD, eigen, LU) — used by #211
cuRAND bindings + IPhiloxGenerator<T> GPU impl + Sobol / Scrambled Sobol (feeds #213, #225)
Thrust/CUB wrappers: DeviceReduce, DeviceScan, DeviceSort, DeviceRadixSort,
                     DeviceHistogram, DeviceRunLengthEncode (feeds #210)
NVTX ranges (Nvtx.Range, Nvtx.Mark)
CudaMemory.Snapshot() / Summary() / Stats() / ResetPeak()
Auto-NVTX from LazyTensorScope + TapeStepContext
```

### HIP / ROCm
```
MIOpen RNN bindings
rocSPARSE / rocSOLVER / rocRAND wrappers
ROCm Thrust equivalent
```

### Metal
```
MPS matrix decompositions (Cholesky/QR/SVD) — feeds #211
Metal RNG
Metal Performance Shaders for vision primitives (informs #217)
```

### OpenCL / Vulkan / WebGPU
```
Device-wide reduce, scan, sort kernels (baseline primitives)
Device RNG (xorshift-derived where no vendor lib exists)
```

### Unified surface
```
IDeviceRng<T> abstraction with backend-specific implementations
IDevicePrimitives<T> abstraction for reduce/scan/sort
```

**Total: ~45 bindings + unified abstractions.**

## 3. Win vectors

1. **cuDNN RNN with symbolic backward**: `SymbolicBackwardGraph` simplifies LSTM state gradients at compile time. PyTorch numerically evaluates.
2. **cuSPARSE SpMM driven by `AutotuneCache`**: measures tipping point between dense GEMM and SpMM.
3. **cuRAND as `IPhiloxGenerator<T>`**: identical interface to CPU Philox — deterministic seeding CPU↔GPU. PyTorch CPU ≠ GPU generators.
4. **Thrust/CUB as `DevicePrimitives.cs`** exposed at tensor level — users compose custom CUDA kernels with our warp-friendly primitives.
5. **NVTX ranges auto-emitted** from `TapeStepContext` + `LazyTensorScope` — zero annotation.
6. **Memory snapshot includes `TensorArena` + `GpuWorkspace`** — unified view. PyTorch covers caching allocator only.
7. **Multi-backend equivalents**: rocSPARSE/SOLVER/RAND for HIP; MPS for Metal; Vulkan/OpenCL/WebGPU scan/reduce/sort kernels.
8. **Persistent RNN kernels** from cuDNN with correctness tests — PyTorch disabled by default due to bugs.

## 4. Per-axis beat targets

| Axis | Target |
|---|---|
| cuDNN LSTM forward/backward vs PyTorch | ≥ parity; symbolic-bwd simplification demonstrably faster |
| cuSPARSE SpMM correctness at 90%+ sparsity | numerical parity vs dense ref |
| cuRAND determinism CPU↔GPU | identical bits for same (seed, subseq, offset) |
| Memory snapshot | lists `TensorArena` + `GpuWorkspace` allocs with sizes |
| NVTX ranges in Nsight | visible without user annotation |
| SpMM auto-dispatch on sparse Transformer | beats always-dense path |
| Multi-backend primitives | MIOpen RNN on HIP; MPS Cholesky on Metal; Vulkan/OpenCL/WebGPU reduce/scan/sort work |

## 5. File inventory

**New (~70):**
- `Engines/DirectGpu/CUDA/CuDnnRnn*.cs`, `CuSparseNative.cs` + `CuSparse.cs`, `CuSolverNative.cs` + `CuSolver.cs`, `CuRandNative.cs` + `CuRand.cs`, `DevicePrimitives.cs`, `Nvtx.cs` (~20)
- `Engines/DirectGpu/HIP/` equivalents — MIOpenRnn*, RocSparse*, RocSolver*, RocRand*, RocmDevicePrimitives, RocmNvtx (~12)
- `Engines/DirectGpu/Metal/` — MpsDecomps, MetalRng (~4)
- `Engines/DirectGpu/OpenCL/`, `Vulkan/`, `WebGPU/` — reduce/scan/sort kernels + xorshift RNG (~18)
- `src/AiDotNet.Tensors/Abstractions/` — IDeviceRng, IDevicePrimitives (~4)

**Modified (~8):** `NativeMemoryManager.cs`, `GpuWorkspace.cs`, `TensorArena.cs`, `LazyTensorScope.cs` (NVTX auto-emit), `TapeStepContext.cs` (NVTX auto-emit), `CompiledInferencePlan.cs`, `CompiledTrainingPlan.cs`.

**Tests (~30):** cuDNN RNN train-loop LM mini-benchmark, cuSPARSE SpMM correctness, cuRAND determinism (CPU↔CUDA bits), memory snapshot assertions, NVTX range visibility, multi-backend primitive parity.

**Benchmarks (~5):** `CuDnnRnnBenchmarks`, `CuSparseBenchmarks`, `CuRandBenchmarks`, `DevicePrimitivesBenchmarks`, `MemorySnapshotBenchmarks`.

## 6. Acceptance gate

1. cuDNN RNN LSTM train-loop on LM mini-benchmark matches CPU ref within tolerance.
2. cuSPARSE SpMM correctness at 90%+ sparsity.
3. cuRAND determinism: same (seed, subseq, offset) → identical bits CPU Philox ↔ cuRAND Philox.
4. Memory snapshot lists all allocator state correctly.
5. NVTX ranges visible in Nsight Systems without annotation.
6. Multi-backend primitives on HIP/Metal/Vulkan/OpenCL/WebGPU ship or skip-with-reason.
7. Persistent cuDNN RNN kernels pass correctness test.

## 7. Timeline — ~8 weeks

W1 cuDNN RNN/LSTM/GRU bindings.
W2 cuSPARSE bindings + SpMM/SpGEMM.
W3 cuSOLVER bindings (shared with #211).
W4 cuRAND bindings + IPhiloxGenerator GPU + Sobol/Scrambled Sobol.
W5 Thrust/CUB wrappers via DevicePrimitives.
W6 NVTX + memory snapshot + auto-emit.
W7 HIP equivalents (MIOpen, rocSPARSE/SOLVER/RAND).
W8 Metal/Vulkan/OpenCL/WebGPU primitives + benchmarks + PR.

## 8. Dependencies

- **Blocks**: #211 (cuSOLVER tight coupling — co-develop), #213 (cuRAND for GPU Philox), #221 (cuSPARSE tight coupling), #223 (cuDNN RNN for CTC loss GPU path), #220 (NVTX for profiler), #225 (cuRAND for graph-safe RNG, NVTX for observability).
- **Blocked by**: nothing.

## 9. Risks

- **Thrust/CUB C++ interop** — small C++ shim DLL required; document build toolchain.
- **Persistent RNN kernels** — PyTorch disabled them for a reason; replicate their bug-test suite.
- **6-backend primitive parity** — Metal/Vulkan/OpenCL/WebGPU RNG quality varies; document per-backend guarantees.

---

*Part of epic #209. Plan ratified 2026-04-19.*

**Comment by ooples** [comment]
## Cross-references from PR #265 (Issue #216 — Data loading)

The following data-loading items were deferred from #216 because they need GPU-side primitives that live in this issue. Adding here so they're tracked once the GPU primitives stack lands:

- [ ] **GPU-direct prefetch for `DataLoader`**: PR #265 ships the host-side async prefetch (Channel-based reorder pipeline). The "decode on CPU, augment on GPU, `pin_memory=True` auto-managed" path from #216's "How we beat PyTorch" point #2 needs the GPU stream pool + pinned-memory allocator that this issue introduces.
- [ ] **`pin_memory`/`pin_memory_device` keyword on `DataLoaderOptions`**: depends on the cuMemHostAlloc-style pinned allocator wrapper (NativeMemoryManager extension) that #219 brings in.

## Cross-references from PR #264 (Issue #220 — Profiler & debug)

PR #264 ships the profiler core (chrome-trace, MemoryProfiler, NVTX/ITT bridges, anomaly localization, bottleneck CLI) but explicitly defers the GPU-event-source wiring that this issue gates:

- [ ] **GPU-side activity capture for `Profiler`**: `ProfilerActivities.Gpu` is wired into the session option flag and `NvtxBridge.PushRange` fires when the activity is enabled, but the profiler doesn't yet capture cuDNN/CUPTI events, kernel launch / H2D / D2H copies — those need this issue's CUPTI / NVTX bridge to be live.
- [ ] **Memory snapshot across the CUDA caching allocator**: PR #264's `MemoryProfiler` covers `TensorAllocator` and the in-process arena. Cross-allocator coverage (`GpuWorkspace`, native CUDA caching allocator, ROCm-equivalent) needs the GPU memory primitive surface.

References: https://github.com/ooples/AiDotNet.Tensors/pull/264 and https://github.com/ooples/AiDotNet.Tensors/pull/265

**Comment by ooples** [comment]
## Cross-references from PR #266 (Issue #215 — Distributed training stack)

PR #266 ships the full distributed framework (ProcessGroup / DDP / FSDP / DTensor / Pipeline / RPC / ElasticLauncher) but defers the GPU-side network transport that this issue gates:

- [ ] **NCCL transport binding for `IProcessGroup`**: PR #266 ships the `InProcessGroup` reference backend that exercises the full collective surface in-process. The GPU-collective NCCL backend (`AllReduce` / `AllGather` / etc. via NCCL primitives, `ncclComm` lifecycle, CUDA stream integration) needs the GPU primitive scaffolding this issue introduces.
- [ ] **GPU-aware FSDP shard placement**: `FullyShardedDataParallel<T>`'s `ParamDtype` mixed-precision option and `CpuOffload` host↔device migration both depend on the GPU memory primitives stack.

References: https://github.com/ooples/AiDotNet.Tensors/pull/266
